### PR TITLE
DrivAerML: Conservative LR neighborhood sweep (lr=3.5e-4/4e-4/4.5e-4)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Every DrivAerML experiment that raised LR above 5e-4 crashed catastrophically (lr=7e-4 → 15.352%). Three independent experiments confirmed lr=3e-4 is worse. This suggests a very narrow LR optimum. We probe the neighborhood just below 5e-4 with minimal perturbations (4e-4, 4.5e-4, 3.5e-4) to determine whether the baseline can be improved with a small, careful LR reduction — or whether 5e-4 truly is the global optimum in this direction.

## Baseline
val_primary/surface_rel_l2_pct = **4.619%** (PR #2691, 4L/512d/8H, T_max=30, lr=5e-4, AdamW, WD=1e-2)

## Runs

All runs keep golden config values: 4L/512d/8H, WD=1e-2, gc=1.0, T_max=30, AdamW, no-use-ema, enable-fourier, 50k surface points.

**Run 1 (CUDA 0): lr=4e-4**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 4e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-lr-conservative --wandb-name drivaerml-lr4e4
```

**Run 2 (CUDA 1): lr=4.5e-4**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 4.5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-lr-conservative --wandb-name drivaerml-lr45e4
```

**Run 3 (CUDA 2): lr=3.5e-4**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 3.5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-lr-conservative --wandb-name drivaerml-lr35e4
```

## Expected Outcome
val_primary/surface_rel_l2_pct < **4.619%**

Report the best val_primary/surface_rel_l2_pct achieved for each run and whether training was stable throughout.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT rei:

## Results

### LR Neighborhood Sweep: lr=3.5e-4, 4e-4, 4.5e-4

**Baseline**: `val_primary/surface_rel_l2_pct = 4.619%` (PR #2691, lr=5e-4, 4L/512d/8H, T_max=30, AdamW, WD=1e-2)

| LR | Best Val (%) | Best Epoch | Total Epochs | Stability | W&B Run ID |
|---|---|---|---|---|---|
| **3.5e-4** | **5.48** | ep185 | 193+ (running) | Most stable (max grad_norm 9.71, no late spikes) | `axku30mp` |
| **4e-4** | **5.53** | ep190 | 194+ (running) | Stable (max grad_norm 10.16, one minor spike at ep161 → 10.1%) | `r2gxs7vf` |
| **4.5e-4** | 15.83 | ep35 | 198+ (running, diverged) | **DIVERGED at ep37** — catastrophic gradient explosion | `wtc7k40v` |

All configs: 4L/512d/8H, cosine T_max=30, gc=1.0, WD=1e-2, AdamW, no-EMA, Fourier features, 50k surface points, batch_size=1, SENPAI_TIMEOUT_MINUTES=180

### Peak Memory
~6,017 MiB per GPU (GPUs 0/1/2)

### Divergence Analysis (lr=4.5e-4)
- Ep30-36: Already oscillating at 15-21% with grad_norm ~1.3 and ~240 clip events/epoch
- **Ep37**: val spikes to 28.5%, grad_norm jumps to 1.62
- **Ep42**: catastrophic blow-up — val 50.9%, grad_norm 6.6
- **Ep43+**: grad_norm reaches Infinity, 394/394 batches clipped every epoch
- Val stabilizes at ~66% (effectively random) and never recovers

### Commands Used
```bash
# GPU 0: lr=4e-4
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 4e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-lr-conservative --wandb-name drivaerml-lr4e4 --agent rei

# GPU 1: lr=4.5e-4
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 4.5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-lr-conservative --wandb-name drivaerml-lr45e4 --agent rei

# GPU 2: lr=3.5e-4
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 3.5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-lr-conservative --wandb-name drivaerml-lr35e4 --agent rei
```

### What Happened

**Neither lr=3.5e-4 nor lr=4e-4 beats the baseline.** Both converge to a similar performance band (5.5-6.0% typical, with stochastic dips to ~5.48% at cosine troughs) — roughly 19% worse than the baseline's 4.619%. The loss landscape appears to have a sharp performance cliff between 4e-4 and 4.5e-4 where stability degrades catastrophically, but the optimal basin near 5e-4 cannot be reached from below.

Key observations:
1. **lr=3.5e-4 and lr=4e-4 are nearly identical** in final performance (5.48% vs 5.53%), suggesting the loss landscape is relatively flat in this region. Lower LR buys marginally more stability but not better minima.
2. **lr=4.5e-4 diverges catastrophically** at ep37, confirming there is a sharp instability boundary. However, this boundary is well below the baseline lr=5e-4, which was stable — suggesting the baseline run may have benefited from specific initialization or scheduling conditions that allowed it to navigate past this region.
3. **The 4.619% baseline appears to occupy a stochastic deep basin** that lower LRs cannot reach. The cosine schedule at lr=5e-4 likely provides enough momentum to explore deeper minima that lr ≤ 4e-4 simply cannot access.

### Suggested Follow-ups
1. **Investigate why lr=5e-4 works but lr=4.5e-4 doesn't**: The instability cliff between 4e-4 and 5e-4 is puzzling. The baseline at 5e-4 is *above* the instability threshold but somehow stable — could be initialization-dependent or schedule-dependent.
2. **Warmup + higher LR**: Try lr=5e-4 or even 5.5e-4 with a longer warmup phase (e.g., 10-20 epochs linear warmup) to see if warmup can stabilize higher LRs that would otherwise diverge.
3. **Gradient clipping threshold sweep**: The current gc=1.0 may be too permissive for lr=4.5e-4. A tighter gc (e.g., 0.5) combined with lr=4.5e-4 might prevent the divergence cascade.
4. **Cosine T_max exploration**: T_max=30 with lr=4e-4 may be too short a period — the model never fully anneals before being kicked back up. A longer T_max (60-90) at lr=4e-4 could allow deeper convergence within each cycle.

**Comment by tcapelle** [comment]
STUDENT rei:

## CRITICAL UPDATE: lr=4e-4 has also diverged (ep202)

**The lr=4e-4 run experienced late-stage catastrophic divergence starting at ep202**, following the same cascade pattern as lr=4.5e-4 but delayed by ~165 epochs.

### Divergence timeline (lr=4e-4):
| Epoch | Val (%) | Grad Norm | Clip Events | LR |
|---|---|---|---|---|
| 201 | 7.12 | 0.24 | 0 | 3.62e-4 |
| 202 | 12.19 | 1.65 | 136 | 4.37e-6 (cosine trough) |
| 203 | 16.96 | 14.77 | 371 | 3.96e-4 (cosine peak) |
| 204 | 10.87 | 13.78 | 383 | 3.82e-5 (cosine trough) |
| **205** | **56.26** | **69.65** | 393 | 3.00e-4 (cosine restart) |
| **206** | **46.08** | **109.39** | 394 | 1.79e-4 |

The cosine restart at ep205 (lr jump from trough to mid-cycle) triggered the catastrophic blowup. The model was already in an unstable state with persistently high grad norms after ep202, and the lr kick back up sealed the divergence.

### Updated results table:
| LR | Best Val (%) | Best Epoch | Status | W&B Run ID |
|---|---|---|---|---|
| **3.5e-4** | **5.33** | ep204 | **Only stable run** — still improving, new best at ep204 | `axku30mp` |
| **4e-4** | 5.53 | ep190 | **DIVERGED at ep202** — late-stage catastrophic instability | `r2gxs7vf` |
| **4.5e-4** | 15.83 | ep35 | **DIVERGED at ep37** — early catastrophic instability | `wtc7k40v` |

### Revised analysis

This changes the narrative significantly:
1. **2 out of 3 LR variants diverged.** The instability boundary extends well below 4.5e-4 — even 4e-4 is ultimately unstable, it just takes 5x longer to hit a critical event.
2. **lr=3.5e-4 is the only truly stable variant** in this sweep, and it just hit a new best of 5.33% at ep204 — still improving.
3. **The baseline lr=5e-4 achieving 4.619% is remarkable.** It sits above BOTH divergence points (4e-4 and 4.5e-4) yet remained stable for its full run. This strongly suggests the baseline benefited from either a fortunate initialization or conditions that this sweep couldn't replicate.
4. **Cosine restart is the divergence trigger.** In both cases (ep37 for lr=4.5e-4, ep205 for lr=4e-4), the catastrophic blowup occurred when the cosine schedule kicked lr back up from a trough while the model was in a marginally unstable state. The lr restart amplifies accumulated instability into a full divergence cascade.

### Revised suggested follow-ups
1. **Longer lr=3.5e-4 run**: This run is still improving at ep204 (5.33%). A longer training window could push it closer to or below the baseline.
2. **Cosine schedule with decay**: Instead of periodic cosine restart, use cosine annealing with decreasing peak LR each cycle (warm restart with decay) to prevent late-stage divergence.
3. **Investigation into baseline stability at lr=5e-4**: What allows lr=5e-4 to be stable when lr=4e-4 and lr=4.5e-4 both diverge? Seed/initialization analysis could be revealing.

**Comment by tcapelle** [comment]
STUDENT rei:

## Final Results (all runs complete)

### Test Metrics
| LR | Best Val (%) | Best Val Epoch | Total Epochs | Test (%) | Status | W&B Run ID |
|---|---|---|---|---|---|---|
| **3.5e-4** | **5.122** | ep234 | 238 | **6.640** | Stable throughout | `axku30mp` |
| **4e-4** | 5.486 | ep199 | 239 | 70.879 | Diverged at ep202 | `r2gxs7vf` |
| **4.5e-4** | 15.825 | ep35 | 243 | 67.985 | Diverged at ep37 | `wtc7k40v` |

**Baseline**: val_primary/surface_rel_l2_pct = 4.619% (lr=5e-4)

### Key takeaways
- **lr=3.5e-4 is the only stable variant** — best val 5.122% (ep234), test 6.640%. Neither beats baseline.
- **Both lr=4e-4 and lr=4.5e-4 diverged catastrophically** — test metrics (~68-71%) reflect permanently broken models.
- **lr=3.5e-4 was still improving** at timeout — its best val dropped from 5.48% → 5.33% → 5.23% → 5.16% → 5.12% over the last ~30 epochs. A longer run could narrow the gap further but is unlikely to reach 4.619%.
- **Conclusion: lr=5e-4 remains the optimal LR for DrivAerML with this config.** Lower LRs are either too slow to converge (3.5e-4) or unstable (4e-4, 4.5e-4).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2849 because ADVISOR: CLOSE — Conservative LR conclusive. lr=3.5e-4 (5.12%), lr=4e-4 (5.49%, diverged), lr=4.5e-4 (15.8%, diverged). lr=5e-4 confirmed optimal. gc=1.0+LR variants being tested in new wave..
